Make "useful window widths" and "useful window heights" user-configurable through PaperWM extension settings

jtaala opened the pull request:
Resolves #316, resolves #402 

This PR makes the "useful window widths" and "useful window heights" user-settable through the extension pref settings ui.

Inputs supported are percentage or pixel values, e.g. `38.195%; 50%; 61.804%` or `400px; 800px; 900px`.  Mixed input types (e.g. input with `%` AND `px` in it) are not supported and will cause the input box to turn red (error css class), as will any invalid input:
![image](https://user-images.githubusercontent.com/30424662/209928988-f7ec9fb9-fc42-43f1-b2e1-fee4f939cce3.png)
![image](https://user-images.githubusercontent.com/30424662/209928895-db071c04-5cf9-4c1e-b780-b76032fe1cdd.png)
![image](https://user-images.githubusercontent.com/30424662/209928938-e45353b4-b731-467c-bf95-d6f340bd8a07.png)

Changes are applied immediately (no need to logout/login etc.).

Reset buttons are also provided which will reset width/height values back to the default PaperWM cycle values.

Feedback welcome!

![image](https://user-images.githubusercontent.com/30424662/209930853-013995a6-7b16-444f-9589-35f4b498aa85.png)

jtaala committed:
Added "useful cycle widths" and "useful cycle heights" to Settings.ui.
Bound prefs cycle-width-steps in prefs.js.

Added reset buttons.

Added reset button tool tips.

jtaala committed:
Updated tooltips mentioning that values can be ratio or pixel values (but that mixed ratio and pixel values are not supported).
Now splitting width/height input values by whitespace blocks (seems a
bit cleaner).

jtaala commented:
@ccope could you review / test this one when you get a chance?

jtaala commented:
@Gelbana @tadfisher @smichel17  - just FYI, asked ccope to review this one (since he requested it in #402).

Thoughts or any improvements suggestions super welcome!

smichel17 commented on Settings.ui:
```diff
@@ -249,6 +249,110 @@
+              <object class="GtkListBoxRow" id="general_row_4">
+                <property name="activatable">False</property>
+                <property name="focusable">False</property>
+                <property name="tooltip_text" translatable="yes">Comma separated values representing the "useful window widths" that will be cycled through. Values can be either ratio or pixel values. Mixed ratio and pixel values are not supported.</property>
```
> ratio or pixel values

For consistency with #453, I'd suggest only supporting `___px` and `__%`, not plain numbers.

jtaala replied:
yeah, I agree, especially if #453 switching to 'px' or '%'.

Cheers.

smichel17 commented on prefs.js:
```diff
@@ -109,6 +109,36 @@ var SettingsWidget = class SettingsWidget {
+        let widthArray = this._settings.get_value('cycle-width-steps').deep_unpack();
+        cycleWidths.set_text(widthArray.toString().replaceAll(',', ', '));
+        cycleWidths.connect('changed', () => {
+            let varr = cycleWidths.get_text()
+                .replaceAll(',', ' ')
+                .split(/\s+/)
```
> can be entered as comma (or whitespace) separated values

I would be stricter with how you parse this.

1. Require a certain delimiter, rather than allowing the user to mix whitespace and commas. Fine to also allow the other one to be present optionally (`, ` separators are easier for *me* to read so I'd like if they were *supported*), but I would require one or the other as the character we actually split on. I don't think there's any edge cases where this matters right now; it's just a case of "why make parsing harder for ourselves if it's not strictly necessary?"
    - I recommend making the whitespace the optional bit. I don't think it will be relevant here, but making whitespace not be a separator leaves open the _option_ of doing crazy things like `max 50% 1000px` as a value. Not saying we _should_ add anything along those lines, just that it's nice to have the flexibility if needed.

2. I'd use a semicolon as a separator, rather than a comma. In much of Europe, comma is a decimal place separator, so it's something we might someday want to support (`33,33%`).

```suggestion
        cycleWidths.connect('changed', () => {
            let varr = cycleWidths.get_text()
                .split(';')
                .map((s) => s.trim())
```

Note: this is (obviously) not the only change that would be needed, just a convenient place to hang this thread.

jtaala replied:
Yep, makes sense.  The original implementation was quite lenient on input format since I only wanted some delimited number array (hence replacing and splitting on whitespace chunks and then mapping to Number worked well).  Given we'll be switching to `%` and `px` strings trimming and processing seems in order.

jtaala replied:
Okay!  requesting a re-review here.  Have implemented all these suggestions.  Parsing is stricter and am enforcing units on each input.  Thanks @smichel17!

jtaala committed:
Parsing processing implements and added some error coloring on the entry
element for invalid inputs.

jtaala committed:
Enforcing unit requirement (either '%' or 'px') at end of each input.  Increased length of input elements.

jtaala committed:
Squashed a corner case to ensure percent inputs <= 100%.  Also added
sort to ensure monotonicity (PaperWM cycle widths/heights require
monotonic values).

jtaala committed:
FIX: cycleProcessor function needs to apply set_value on `settingName` variable.
Was using `cycle-width-steps` instead of the variable `settingName` so `cycle-height-steps` wasn't being applied.

smichel17 commented:
On reflection, I use "useful window widths" a _ton_ but I don't think I've ever actually used "useful window heights" (and in fact, I didn't realize it was a thing until just now). Is it something you added because you wanted to change it, or because you were also making the window widths configurable and it felt like the heights should be, too?

jtaala commented:
> On reflection, I use "useful window widths" a _ton_ but I don't think I've ever actually used "useful window heights" (and in fact, I didn't realize it was a thing until just now). Is it something you added because you wanted to change it, or because you were also making the window widths configurable and it felt like the heights should be, too?

Heights were there, I actually use them quite a bit as I have a layout I like for several applications and being able to set a specific height (that I use) was part of the early intent of this PR.  So bit of a mix - they were there (so might as well have a setting for them like widths), plus I wanted to change/use them anyway.

jtaala commented:
Yep, I guess I wouldn't mind have the width/height settings hidden / or changing them through dconf, a few thoughts though:

- having them in settings allows changing them without needing a shell-reload (which is handy for fine-tuning without needing restart/login-logout if in wayland)... but there's ways to keep that behaviour with changes via gsettings/dconf;
-  it's much easier to let users know when they input invalid inputs if in settings (e.g. input boxes turn red) - users will probably try to input/mix % and px values (or some other invalid inputs) and currently the box turns red letting them know when inputs aren't right.  We could throw an error user notification when they try to use the invalid widths/heights though... so maybe this one isn't such a big deal.
